src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,514 @@ ep_rt_mono_write_event_ee_startup_start (void)
+// max-number-of-descriptors per event, which in the case of BulkType, will kick in
+// far sooner. There's a max number of 128 descriptors allowed per event. 2 are used
+// for Count + ClrInstanceID. Then 4 per batched value. (Might actually be 3 if there
+// are no type parameters to log, but let's overestimate at 4 per value).
+static const uint32_t k_max_count_type_values = (128 - 2) / 4;
+
+struct _BulkTypeEventLogger {
+	BulkTypeValue m_rg_bulk_type_values[k_max_count_type_values];
+	uint32_t m_n_bulk_type_value_count;
+	uint32_t m_n_bulk_type_value_byte_count;
+	uint8_t *m_p_bulk_type_event_buffer;
+};
+
+static
+BulkTypeEventLogger*
+ep_rt_bulk_type_event_logger_init()
+{
+	BulkTypeEventLogger *p_type_logger = g_malloc0 (sizeof(BulkTypeEventLogger));
+	for (int i = 0; i < k_max_count_type_values; i++)
+		ep_rt_bulk_type_value_init (&p_type_logger->m_rg_bulk_type_values[i]);
+	p_type_logger->m_p_bulk_type_event_buffer = g_malloc0 (sizeof(uint8_t) * k_size_of_event_buffer);
+	return p_type_logger;
+}
+
+static
+void
+ep_rt_bulk_type_event_logger_fini(BulkTypeEventLogger *p_type_logger)
+{
+	g_free (p_type_logger->m_p_bulk_type_event_buffer);
+	g_free (p_type_logger);
+}
+
+//---------------------------------------------------------------------------------------
+//
+// ep_rt_mono_fire_bulk_type_event fires an ETW event for all the types batched so far,
+// it then resets the state to start batching new types at the beginning of the
+// m_rg_bulk_type_values array.
+//
+// This follows CoreCLR's BulkTypeEventLogger::FireBulkTypeEvent
+
+void
+ep_rt_mono_fire_bulk_type_event (BulkTypeEventLogger *p_type_logger)
+{
+	if (p_type_logger->m_n_bulk_type_value_count == 0)
+		return;
+
+	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "Firing BulkTypeEvent!");
+
+	uint16_t n_clr_instance_id = clr_instance_get_id();
+
+	uint32_t i_size = 0;
+
+	for (int i_type_data = 0; i_type_data < p_type_logger->m_n_bulk_type_value_count; i_type_data++)
+	{
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "TypeData[%d]\n", i_type_data);
+		BulkTypeValue *target = &p_type_logger->m_rg_bulk_type_values[i_type_data];
+
+		memcpy(p_type_logger->m_p_bulk_type_event_buffer + i_size,
+			&target->fixed_sized_data,
+			25); // sizeof(target->fixed_sized_data));
+		i_size += 25; // sizeof(target->fixed_sized_data);
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "FSB\ntype_id: %d\nmodule_id: %d\ntype_name_id: %d\nflags: %d\ncor_element_type: %d\n", target->fixed_sized_data.type_id, target->fixed_sized_data.module_id, target->fixed_sized_data.type_name_id, target->fixed_sized_data.flags, target->fixed_sized_data.cor_element_type);
+
+		char *wsz_name = target->s_name;
+		if (!wsz_name || strlen(wsz_name) == 0)
+		{
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+			p_type_logger->m_p_bulk_type_event_buffer[i_size++] = 0;
+		}
+		else
+		{
+			uint32_t name_size = (strlen(target->s_name) + 1) * sizeof(wchar_t);
```

Comment:
There shouldn't be any uses of `wchar_t` in Mono. if this is supposed to be UTF16, then use `gunichar2`.

Reply:
Thanks, it is now using `gunichar2` for writing to the bulk type event logger buffer, and converting `s_name` from `ep_char8_t` to `gunichar2` using `g_utf8_to_utf16 (target->s_name, -1, NULL, NULL, NULL)` and freeing the bytes copied as discussed

Reply:
In EventPipe we use ep_char8_t (UTF8) as much as possible. If this string only need to be UTF16 in the serialized event payload (that's how nettrace strings are defined), then the utf8 -> utf16 should already be done inside the EventPipeWriteEvent* generated functions, so you should be able to fire an event with a string as UTF8 and it will be converted by the serializer. For an example, look at FireEtwMethodDCEndVerbose_V1, all parameters to that function is ep_char8_t * and will automatically be converted before written into the payload buffer.

Reply:
For each EventPipe provider there are auto generated functions that does serialization into buffers for all FireEtw* generated functions, there you find functions to serialize most EP types. You could use that as inspiration to align this with how all the other FireEtw* serialize parameters into the payload.